validator/client/src/main/java/tech/pegasys/teku/validator/client/StableSubnetSubscriber.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import static java.lang.Integer.min;
+import static tech.pegasys.teku.util.config.Constants.ATTESTATION_SUBNET_COUNT;
+import static tech.pegasys.teku.util.config.Constants.EPOCHS_PER_RANDOM_SUBNET_SUBSCRIPTION;
+import static tech.pegasys.teku.util.config.Constants.RANDOM_SUBNETS_PER_VALIDATOR;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import tech.pegasys.teku.validator.api.SubnetSubscription;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+
+public class StableSubnetSubscriber {
+
+  private final ValidatorApiChannel validatorApiChannel;
+  private final Set<Integer> availableSubnetIndices =
+      IntStream.range(0, ATTESTATION_SUBNET_COUNT).boxed().collect(Collectors.toSet());
+  private final NavigableSet<SubnetSubscription> subnetSubscriptions =
+      new TreeSet<>(Comparator.comparing(SubnetSubscription::getUnsubscriptionSlot));
+  private final Random rand = new Random();
+
+  private volatile int validatorCount;
+
+  public StableSubnetSubscriber(ValidatorApiChannel validatorApiChannel, int validatorCount) {
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorCount = validatorCount;
+    onSlot(UnsignedLong.ZERO);
+  }
+
+  public void onSlot(UnsignedLong slot) {
+    boolean updated = adjustNumberOfSubscriptionsToNumberOfValidators(slot, validatorCount);
+
+    // Iterate through current subscriptions to replace the ones that have expired
+    final Iterator<SubnetSubscription> iterator = subnetSubscriptions.iterator();
+    while (iterator.hasNext()) {
+      final SubnetSubscription subnetSubscription = iterator.next();
+      if (subnetSubscription.getUnsubscriptionSlot().compareTo(slot) > 0) {
+        break;
+      }
+
+      iterator.remove();
+      int subnetId = subnetSubscription.getSubnetId();
+      availableSubnetIndices.add(subnetId);
+      subscribeToNewRandomSubnet(slot);
+      updated = true;
```

Comment:
This should wind up throwing a `ConcurrentModificationException` because you're adding a new subscription into the map while iterating over it. The unit tests aren't picking it up because they're only replacing one expired subscription so the loop immediately exits after the change. If there was another entry the call to `next()` would fail though.

I'd suggest just iterating through the existing subscriptions, removing any that have expired, then call `adjustNumberOfSubscriptionsToNumberOfValidators` to create as many more subscriptions as are now required.

Reply:
Good catch.

Reply:
I modified the test so that it would catch the exception, and it did. And I made the fix.